test/blockchaintest/blockchaintest_runner.cpp

```diff
@@ -81,18 +84,28 @@ TransitionResult apply_block(TestState& state, evmc::VM& vm, const state::BlockI
     }
 
     std::vector<state::Requests> requests;
+    auto system_contracts_failed = false;
+
     if (rev >= EVMC_PRAGUE)
-        requests.emplace_back(collect_deposit_requests(receipts));
+    {
+        auto opt_deposits = collect_deposit_requests(receipts);
+        if (!opt_deposits.has_value())
+            system_contracts_failed = true;
+        else
+            requests.emplace_back(std::move(*opt_deposits));
+    }
 
-    auto system_call_requests = system_call_block_end(block_state, block, block_hashes, rev, vm);
-    std::ranges::move(system_call_requests, std::back_inserter(requests));
+    auto requests_result = system_call_block_end(block_state, block, block_hashes, rev, vm);
+    system_contracts_failed |= !requests_result.has_value();
+    if (!system_contracts_failed)
```

Comment:
here if deposits collection failed, but `system_call_block_end()` succeeded you still will skip collecting `requests_result`.

Maybe not a big problem, but also maybe makes sense to skip `system_call_block_end()` altogether if deposits collection already failed?

Reply:
Refactored...